### DISCUSSION TITLE
Data demonstration

### DISCUSSION THREAD

**Discussion by mnmneE** [wehr-lab/autopilot#122] [Q&A] [answered]
Hi there,

after implementing the free_water task;

, I tried to show the data but I couldn't find how to do it in the terminal with both the blink and the free water tasks, I am getting an  error related to plot parameters even though I don't need to plot and no instruction in the code to plot any data (please find attached a screenshot for the error);
I followed the documentation where it is mentioned that in order to show the data the user  will be able to access the data from the run with  :meth:`.Subject.get_trial_data`; can you please  provide one example about it; will the data be shown automatically if I don't have the plot error; or I should explicitly  use the method in the terminal ;

yet I am blocked because of the error related to the plot parameters, I tried also to show the data using HDF viewer but it didn't work;

thanks for your help.
regards
![Screenshot from 2021-08-17 14-34-47](https://user-images.githubusercontent.com/45367650/129726674-561bc68c-a4b8-41f7-b122-0567ff5f161e.png)

**Comment by sneakers-the-rat** [comment, marked as answer]
It looks like there are a few errors in that screenshot -- the first two are from not having a subject selected in the subject list in the gui (the list next to the start/stop button) when pressing start/stop button.

the second is from a misspecified `PLOT` dictionary in the task - this is documented here: https://docs.auto-pi-lot.com/en/main/guide/task.html#plot. In this case it's sort of odd because a preceding line checks if the `PLOT` dictionary is empty, which is the typical way to specify a plotless task, and returns quietly. So the dictionary must have something in it, but not have a `data` subdictionary. I just modified the behavior of the code here https://github.com/wehr-lab/autopilot/commit/9a8c0e6baab325872a74a2e96407314cd3fc7898 so that if there is no `data` key it continues quietly. 

It sounds like you are talking about two different things though, the `PLOT` dictionary specifies how the plot drawn in the GUI looks while the task is being performed. In other words, it maps the task data (like the target, whether the subject got the trial correct, etc) to graphical primitives like points and lines. eg. "draw the `correct` value as a `point` at 0 or 1." I haven't tried to robustly support post-collection plotting since pandas and matplotlib cover that better than I could already. 

To get the trial data, yes you would use [Subject.get_trial_data()](https://docs.auto-pi-lot.com/en/main/core/subject.html#autopilot.core.subject.Subject.get_trial_data). That would look like

```python
from autopilot.core.subject import Subject

# instantiate the subject class with the subject id/name of the subject
# it should find the .h5 file automatically, but you can also pass the directory, etc
# see the docs page for more instantiation params
sub = Subject('subject_id')

# return data from the most recent step in the training protocol
last_step = sub.get_trial_data()

# get all data from all steps
all_data = sub.get_trial_data('all')
```

from there you could use [pandas DataFrame plotting methods](https://pandas.pydata.org/pandas-docs/stable/user_guide/visualization.html) or whatever method you choose.

Does that help? lmk if i missed your question :)

**Reply by sneakers-the-rat** [reply]
as an aside, yes the hdf viewer is very buggy for me as well.

**Reply by mnmneE** [reply]
Thanks a lot, surly it is very helpful and informative.

**Reply by mnmneE** [reply]
I tried to import Subject from  autopilot core yet it throws the following error related to task list;

ImportError : cannot import name 'TASK_LIST' from 'autopilot.tasks'  
should I modify and can you please tell me what parameters may solve the issue

In advance thanks very much for your help

**Reply by mnmneE** [reply]
I just commented the import line in the core script and it worked.

**Reply by sneakers-the-rat** [reply]
try updating your autopilot version, the latest version (>0.4.0) does away with the task list concept in favor of a more general registry system, documented here :) https://docs.auto-pi-lot.com/en/latest/guide/plugins.html
